src/docs/sphinx/dev_guide/weak_form.rst

```diff
@@ -0,0 +1,266 @@
+.. _header-n2:
+
+=========
+Weak Form
+=========
+
+``WeakForm`` is a class that is used specify and evaluate
+finite-element-type calculations. For example, the weighted residual for
+a solid mechanics simulation may look something like:
+
+.. math::
+
+   r(u) := 
+   \underbrace{\int_\Omega \nabla\psi : \sigma(\nabla u) \; \text{d}v}_{\text{stress response}}
+   \;+\;
+   \underbrace{\int_\Omega \psi \cdot b(\mathbf{x}) \; \text{d}v}_{\text{body forces}} 
+   \;+\;
+   \underbrace{\int_{\partial\Omega} \psi \cdot \mathbf{t}(\mathbf{x}) \; \text{d}a}_{\text{surface loads}},
+
+where :math:`\psi` are the test basis functions. To describe this
+residual using ``WeakForm``, we first create the object itself, providing a
+template parameter that expresses the test and trial spaces (i.e. the
+"inputs" and "outputs" of the residual function, :math:`r`). In this
+case, solid mechanics uses H1 test and trial spaces, so we write:
+
+.. code-block:: cpp
+
+   constexpr int order = 2; // the polynomial order of the basis functions
+   constexpr int dim = 3; // the dimension of our problem
+   using test = H1<order, dim>;
+   using trial = H1<order, dim>;
+   WeakForm< test(trial) > residual(test_fes, trial_fes);
+
+where ``test_fes``, ``trial_fes`` are the ``mfem`` finite element spaces for the problem. 
+The template argument follows the same convention of ``std::function``:
+the output-type appears outside the parentheses, and the input-type(s)
+appear, inside the parentheses (in order). So, the last line of code in
+the snippet above is saying that ``residual`` is going to represent a
+calculation that takes in an H1 field (displacements), and returns a
+vector of weighted residuals, using H1 test functions.
+
+Now that the ``WeakForm`` object is created, we can use the
+following functions to define integral terms (depending on their
+dimensionality). Here, we use :math:`s` to denote the "source" term
+(integrated against test functions), and :math:`f` to denote the 
+"flux" term (integrated against test function gradients).
+
+1. Integrals of the form:
+   :math:`\displaystyle \iint_\Omega \psi \cdot s + \nabla \psi : f \; da`
+
+   .. code-block:: cpp
+
+      WeakForm::AddAreaIntegral([](auto ... args){
+      	auto s = ...;
+      	auto f = ...;
+      	return std::tuple{s, f};
+      }, domain_of_integration);
+
+2. Integrals of the form:
+   :math:`\displaystyle \iiint_\Omega \psi \cdot s + \nabla \psi : f \; dv`
+
+   .. code-block:: cpp
+
+      WeakForm::AddVolumeIntegral([](auto ... args){
+      	auto s = ...;
+      	auto f = ...;
+      	return std::tuple{s, f};
+      }, domain_of_integration);
+
+3. Integrals of the form:
+   :math:`\displaystyle \iint_{\partial \Omega} \psi \cdot s \; da`
+
+   .. code-block:: cpp
+
+      WeakForm::AddSurfaceIntegral([](auto ... args){
+      	auto s = ...;
+      	return s;
+      }, domain_of_integration);	
+
+So, for our problem (since we assumed 3D earlier) we can make an
```

Comment:
Should an example of `Weakform::AddDomainIntegral(...)` be here as well? Also can you add something a bit more helpful that just repeating "Integrals of the form"?